README.md
binfunest Release
================
Phil Shea
2022-10-25

# `Introduction`

<!-- badges: start -->

[![R-CMD-check](https://github.com/PhilShea/binfunest/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/PhilShea/binfunest/actions/workflows/R-CMD-check.yaml)
[![Codecov test
coverage](https://codecov.io/gh/PhilShea/binfunest/branch/main/graph/badge.svg)](https://app.codecov.io/gh/PhilShea/binfunest?branch=main)
[![CRAN
status](https://www.r-pkg.org/badges/version/binfunest)](https://CRAN.R-project.org/package=binfunest)
<!-- badges: end -->

The goal of `binfunest` is to simplify the estimation of offsets and
Back-to-Back “Q” for communications systems given measurements of bit
errors vs. SNR. The package also provides theoretical performance
equations for many common modulation schemes, and eases maximum
likelihood estimation of any function generating binomial probabilities.

## Installation

`binfunest` is available on CRAN, so `install.packages( "binfunest")`
will install it, and `library( binfunest)` will make it available. You
can install the development version of `binfunest` from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("PhilShea/binfunest")
```

## Example

This is a basic example which shows how to solve a common problem of
estimating the B2B Q and offset. First we create a function which will
add offset and B2B parameters to a standard communications modulation
(Quadrature Phase Shift Keying, or QPSK). The package provides
theoretical performance curves for many common modulations. Then
`rbinom` is called to simulate measurements made at a range of signal to
noise ratios (SNR). All SNRs, offsets, and B2BQs are in Decibels. Note
that the generated samples include zeros, and that the first of these
(i.e., the lowest SNR that produced zero errors) is useful in the
parameter estimate.

``` r
require( binfunest)
require( stats4)
#> Loading required package: stats4
QPSKdB.B2B <- B2BConvert( QPSKdB)
O1 <- 3 # offset
B1 <- 15 # B2BQ
s <- 0:25 # SNR Range 
N <- 1000000 # Number of samples
(r <- rbinom( length( s), N, QPSKdB.B2B( s, B1, O1)))
#>  [1] 161958 135306 109002  85035  63269  44868  29705  18587  10921   5923
#> [11]   2749   1382    593    234     88     34     12      4      3      0
#> [21]      0      1      0      0      0      0
df <- data.frame( Errors=r, SNR=s, N=N) # place data in data frame

## This shows how you could do the work by hand
llsb2 <- function( b2b, offset)
       -sum( dbinom( r, N, QPSKdB.B2B( s, b2b, offset), log=TRUE))
mle1 <- stats4::mle( llsb2, start=c( b2b=20, offset=0), nobs=length(s),
                   method="Nelder-Mead")
stats4::coef( mle1)
#>      b2b   offset 
#> 14.96665  2.99366


# Below is the new function
est1 <-  mleB2B( data=df, Errors="Errors", N=N, f=QPSKdB.B2B,
                 fparms=list( x="SNR"), start=c(b2b=20, offset=0))

(est1coef <- stats4::coef( est1))
#>      b2b   offset 
#> 14.96665  2.99366
```

The plot below compares the theoretical curve to the curve with the
chosen B2B and offset, and the curve with the estimated parameters. Note
that all samples above 18 dB SNR returned zero, except the 21 dB point.
The Estimated line is dashed so the 3 dB + Offset line, which it
overlays, is visible.

<img src="man/figures/README-plotQab-1.png" width="100%" />


NEWS.md
# binfunest 0.2.0

* Added the generic $Q$ function vignette demonstrating estimating very high B2B Q.
* Added `cor.mle` function to return the correlation matrix of the `mle` fit.
* Added `DBPSKdB` function.
* Added `marcumq` function (essentially copied from `gsignal` package).
* Corrected the `DQPDKdB` function (and why `marcumq` was needed).
* Simplified `B2BConvert`, and improved retention of function call in calls to `stats4::mle`.

# binfunest 0.1.0

* Added a NEWS.md file to track changes to the package.
* This is the first release of the package.


tests/testthat.R
library(testthat)
library(binfunest)

test_check("binfunest")


tests/testthat/test-cor.R
test_that("cor.mle works", {
   require("stats4", quietly = TRUE)
   Q_ab <- function( gamma, a, b)
      Q_(sqrt( a * gamma / (1 + b * gamma)))
   nobs <- 6 # This is the Mathematica test data.
   s <- c( sqrt(10), 10, 10 * sqrt(10), 100, 100 * sqrt(10), 1000)
   r <- c( 246, 211, 220, 23, 9, 6)
   N <- c( 20000, 500000, 20000000, 20000000, 20000000, 20000000)
   m <-  mleB2B( Errors=r, N=N, f=Q_ab, fparms=list( gamma=s),
                 start=c(a=2, b=undB(-11)), method="L-BFGS-B", lower=list(b=0))
  expect_equal( cor.mle( m), array( c(1., 0.9608065, 0.9608065, 1.), c(2,2)))
})

test_that("cor.mle failos when it should", {
   x <- 1:10 # create a simple vector to pass
   expect_error( cor.mle( x))
})


tests/testthat/test-mleB2B.R
test_that("mleB2B main function works", {
   require("stats4", quietly = TRUE)
   # Set up test data.
   QPSKdB.B2B <- B2BConvert( QPSKdB)
   O1 <- 3
   B1 <- 16
   s  <- 0:20
   N  <- 1000000
   r  <- rbinom( length( s), N, QPSKdB.B2B( s, B1, O1))
   df <- data.frame( Errors=r, SNR=s, N=N)

   llsb2 <- function( b2b, offset)
      -sum( dbinom( r, N, QPSKdB.B2B( s, b2b, offset), log=TRUE))

   mle1 <- stats4::mle( llsb2, start=c( b2b=20, offset=0), nobs=length(s),
                        method="Nelder-Mead")

   # tests using dataframe for errors and snr, singular N.
   est1 <-  mleB2B( data=df, Errors="Errors", N=N, f=QPSKdB.B2B,
                    fparms=list( x="SNR"), start=c(b2b=20, offset=0))
   expect_type( est1, typeof( mle1))
   expect_equal( coef( mle1), coef( est1))

   #tests vector errors, snr and N.
   est2 <-  mleB2B( Errors=r, N=df$N, f=QPSKdB.B2B, fparms=list( x=s),
                    start=c(b2b=20, offset=0))
   expect_equal( coef( mle1),  coef( est2))

   # test N as a name.
   est3 <-  mleB2B( data=df, Errors="Errors", N="N", f=QPSKdB.B2B,
                    fparms=list( x="SNR"), start=c(b2b=20, offset=0))
   expect_equal( coef( mle1),  coef( est3))

   # test use of anonymous function.
   est4 <- mleB2B( data=df, Errors="Errors", N="N",
                   f=function(x, b2b = 20, offset = 0)
                      QPSKdB(-dB(( undB(-s) + undB( -b2b))) - offset),
                   fparms=list( x="SNR"), start=c(b2b=20, offset=0))
   expect_equal( coef( mle1),  coef( est4))
})

test_that( "mleB2B fails when it should", {
   # Set up test data.
   QPSKdB.B2B <- B2BConvert( QPSKdB)
   O1 <- 3
   B1 <- 16
   s <- 0:20
   N <- 1000000
   r <- rbinom( length( s), N, QPSKdB.B2B( s, B1, O1))
   df <- data.frame( Errors=r, SNR=s, N=N)

   expect_error( mleB2B( data=df, Errors="Err", N=N, f=QPSKdB.B2B,
                         fparms=list( x="SNR"), start=c(b2b=20, offset=0)),
                 "Err is not in data")
   expect_error( mleB2B( data=df, Errors="Errors", N="n", f=QPSKdB.B2B,
                         fparms=list( x="SNR"), start=c(b2b=20, offset=0)),
                 "n is not in data")
   #errors in fparm will generate odd messages.
   expect_error( mleB2B( data=df, Errors="Errors", N="N", f=QPSKdB.B2B,
                         fparms=list( x="S"), start=c(b2b=20, offset=0)))
})

test_that( "mleB2B method, control, and '...' work", {
   require("stats4", quietly = TRUE)
   # Set up test data different from above.
   QPSKdB.B2B <- B2BConvert( QPSKdB)
   O2 <- 3
   B2 <- 80
   s <- 0:20
   N <- 1000000
   r2 <- rbinom( length( s), N, QPSKdB.B2B( s, B2, O2))
   df2 <- data.frame( Errors=r2, SNR=s, N=N)

   llsb2 <- function( b2b, offset)
      -sum( dbinom( r2, N, QPSKdB.B2B( s, b2b, offset), log=TRUE))
   mle1 <- stats4::mle( llsb2, start=c( b2b=20, offset=0), nobs=length(s),
                        method="Brent", fixed=list(b2b=80), lower=c( 0, -6),
                        upper=c( 100, 10))
   est1 <-  mleB2B( data=df2, Errors="Errors", N=N, f=QPSKdB.B2B,
                    fparms=list( x=s), start=c(b2b=20, offset=0),
                    method="Brent", fixed=list(b2b=80), lower=c(0, -6),
                    upper=c( 100, 10))
   expect_equal( coef( mle1), coef( est1))
})


R/Data.R
#' An example `BERDF` dataframe created by `simsigs()`, a function in a
#' forthcoming package `coherent`.
#'
#' `BERDF` is a standard `R` data frame created by the `simsigs()` function in
#' the forthcoming `coherent` package.  The observations have been `condense`d
#'
#' @format A dataframe with the following fields:
#' \describe{
#'     \item{Name}{Name of constellation used to create the record.}
#'     \item{SNR}{The SNR in Decibles of the observation.}
#'     \item{Bps}{The number of bits per symbol.  The number of bits in a
#'     simulation run is \code{Bps * N}}
#'     \item{NoisePower}{The actual noise power in the simulation run.  Since
#'     the noise is randomly generated, this is a stochastic item.}
#'     \item{N}{The number of symbols in the simulation run.}
#'     \item{SER}{The number of symbols errors observed in the simulation run.}
#'     \item{BER}{The number of bit errors observed in the simulation run.}
#' }
"BERDFc"


R/mleB2B.R
#' mleB2B  Estimates Back-to-Back "Q" and Offsets to a bit error rate function.
#'
#' Bit error counts modeled as independent binary decisions result in a
#' log-likelihood dependent on the bit error probability.  This function
#' inserts the supplied bit error probability function into the binomial
#' log-likelihood function, and passes that to [stats4::mle], which ultimately
#' calls [stats::optim].  The function will optimize a binomial probability
#' of the form $r = N * P( x_1, x_2, ..., x_n, a_1, a_2, ... a_k)$, where the
#' $x_i$ are variables from `data`, and the $a_j$ are parameters to be
#' estimated.
#'
#' The function estimates the parameters identified in `start` in the
#' constructed call to `f`. For a function `f` of the form
#' `fun( SNR, x2, x3, B2B, offset)`
#' A call of the form
#'
#' \code{
#' mleB2B( data=df, Errors="r", N="trials", f=fun,
#'         fparm=list( SNR="s", x2=1, x3="noise"), start=list(B2B=1, offset=2))
#' }
#'
#'  will construct a call to `mle` of the form:
#'
#' \code{
#' mle( minuslogl=ll, start=start, nobs=length( Errors), method=method)}
#'
#' where the function `ll` is defined as
#'
#' \code{
#'   ll <- function( a, b) -sum( dbinom( df$r, df$n,
#'                 fun( SNR=df$s, x2=1, x3=df$noise, B2B=B2B, offset=offset),
#'                 log=TRUE))
#' }
#'
#' @param data a data frame or list with named components. If a list, each
#' component must be the same length (just like a data frame). This is not
#' checked, so usual rules of recycling will apply.  Partial matching not
#' performed, so you must use full column names.
#' @param Errors A vector of error counts, or a string identifying a column of
#'     `data` from which to draw the error counts
#' @param N A single number, or a vector of the same length as
#'    `data`, or a string identifying a column of `data` specifying the number
#'    of trials used to measure the error counts
#'    in `Errors`.  If a single number, then that number is used as the number
#'    of trials for all error counts.
#' @param f A function that predicts the probability of errors.
#' @param fparms a list of named components that are the arguments of `f`.
#'     Each component can be a string, a single number, or a vector. If a
#'     string that names a column of `data`, that column will be used,
#'     otherwise the string will be passed to `f`.  Note the potential for
#'     errors if a column name was misspelled. A single number or vector
#'     will be passed to `f`. Between `fparms`, `start`, and function defaults,
#'     all parameters that need to be supplied to `f` should be specified, and
#'     (except for defaults) not duplicated.
#' @param start Named list of initial values for the parameters of `f` to be
#'     estimated.
#' @param method Optimization method.  See [stats::optim()].
#' @param ... Optional arguments to be passed to [mle].
#'
#' @return An object of class [stats4::mle-class] with the parameters
#' identified in `start` estimated.
#'
#' @seealso [stats4::mle()], [stats::optim()]
#' @export
#'
#' @examples
#' QPSKdB.B2B <- B2BConvert( QPSKdB)
#' O1 <- 3
#' B1 <- 16
#' s <- 0:20
#' N <- 1000000
#' r <- rbinom( length( s), N, QPSKdB.B2B( s, B1, O1))
#' df <- data.frame( Errors=r, SNR=s, N=N)
#' llsb2 <- function( b2b, offset)
#'          -sum( dbinom( r, N, QPSKdB.B2B( s, b2b, offset), log=TRUE))
#' mle1 <- stats4::mle( llsb2, start=c( b2b=20, offset=0), nobs=length(s),
#'                      method="Nelder-Mead")
#' est1 <-  mleB2B( data=df, Errors="Errors", N=N, f=QPSKdB.B2B,
#'                  fparms=list( x="SNR"), start=c(b2b=20, offset=0))
#'
mleB2B <- function( data=NULL, Errors, N, f, fparms, start,
                    method="Nelder-Mead", ...) {
   cll <- match.call( expand.dots=FALSE)
   if (!is.null( data)) { # a data frame has been supplied.
      dnames <- names( data)
      # first, we need a helper function
      helper <- function( x, strict=FALSE)
         # if in dnames it will pull from data.
         if (is.character( x))
             if (x %in% dnames)  data[[ x]] else
                if (strict) stop( paste( x, "is not in data")) else x
         else x # if (is.character( x))
      Errors <- helper( Errors, strict=TRUE)
      N <- helper( N, strict=TRUE)
      fparms <- lapply( fparms, helper) # expand each item in fparms.
   }
   # start has the names of variables of f to optimize.
   opt_names <- lapply( names( start), as.name)
   # create the call to f
   fun.txt <- c( f, fparms, opt_names)

   fun <- as.call( fun.txt)
   le <- length( Errors)
   if (length( N) == 1) N <- rep( N, le) # normalize N to length of Errors
   ll <- eval( bquote( function()
      -sum( sort(stats::dbinom( Errors, N, .(fun), log=TRUE)))))
   formals( ll) <- start # must use alist to make this work.
   # constructing the call to mle.
   sublst <- list( minuslogl=utils::removeSource( ll), start=start, nobs=le,
                   method=method)
   if (!is.null( cll$...)) sublst <- append( sublst, cll$...)

   #fun1 <- quote( stats4::mle) #get( "mle", asNamespace( "stats4"))
   #do.call( fun1, sublst)
   # This idiom will allow the call to be retained my mle.
   eval( as.call( c( quote( stats4::mle), sublst)), envir=parent.frame())
}


R/theory.R
#' Theoretical error rate functions
#'
#' Functions to calculate the theoretical performance of common modulation
#' formats.  Includes the functions `dB(x)` (returns `10*log10(x)`), `undB(x)`
#' (reverses `dB(x)`), `Q_( x)` (Markum's Q function), and `Q_Inv(x)`
#' (returns the
#' SNR in Decibels to get probability x).  Also includes `mod_Inv`, which returns
#' the SNR required for a the function `f` to reach the supplied BER (bit
#' error rate, or bit error probability).
#'
#' The `marcumq` function is copied from the `gsignal` package, which copied it
#' from the help file of the `lmomco` package, but is clear from Proakis
#' equations 2-1-121 and 2-1-124.
#'
#' The rest of the functions return the probability of a bit error given the
#' SNR in Decibels.
#' *  `QPSKdB` is Quadrature Phase shift keyed: two bits per symbol. Note that
#'             BPSK and QPSK have the same performance when SNR is $E_b/N_0$.
#' *  `DBPSKdB` is differentially detected differential binary PSK.
#' *  `DQPSKdB` is differentially detected differentially coded QPSK.
#' *  `DQPSKDDdB` is differentially decoded differential QPSK (coherently
#'      detected but differentially decoded. See `DQPSK` above.
#' *  `PSQPSKdB` is polarization-shifted QPSK: it is dual pole, but only
#'      one pole is active at any one time, thus supplying three bits per
#'      symbol. (See Agrell & Karlsson (2009, DOI:10.1109/JLT.2009.2029064)).
#' *  `MPSKdB(x, M)` is generic M-ary phase shift keying of `M` points in a circle.
#' *  `MPSKdB.8` simply returns `MPSKdB(x, 8)`
#' *  `QAMdB.8.star` is the optimal star configuration of 8-ary Quadrature
#'      Amplitude Modulation (QAM), such that
#'      the legs are at \eqn{\pm1} and \eqn{\pm(1+\sqrt3)}.
#' *  `QAMdB(x, M)` is generic rectangular QAM constellation of `M` points.
#' *  `QAMdB.16` Returns the BER for the rectangular QAM constellation according to
#'      Proakis Eq. 5-2-80.
#' * `mod_Inv` will take a function `f(x)` and return the x such that
#'      `f(x)==perr`
#'      but it does this based on the `log( f(x))` and the `log( perr)`, so
#'      `f(x)>0`.
#' * `mod_InvV` is a vectorized version (give it a vector of BERs and it returns a
#'      vector of SNRs).
#'
#' @param x The SNR (Eb/No) in Decibels, possibly a vector.
#' @param M The integer number of symbols > 4.
#' @param perr a probability of a bit error.
#' @param f a function (usually a BER function).
#' @param pv a vector of BERs.
#' @param guess a guess for the `perr` (the default usually works).
#' @param offset an offset in Decibels for guesses in `mod_InvV`.
#' @param a,b,m `marcumq` input arguments: non-negative real numbers.
#'
#' @name Theoretical
NULL
#>

sqrt_2 <- sqrt( 2.0)
tms3_ <- 3.0 - sqrt( 3.0) # Need this for 8QAM Star.

#' @rdname Theoretical
#' @name is.wholenumber
#' @param x a real number
#' @param tol the tolerance to test x with.
#' @return `is.wholenumber(x)` returns `TRUE` if `c-round(x) < tol`.
#' @export
is.wholenumber <- function( x, tol = sqrt( .Machine$double.eps))
   abs(x - round( x)) < tol # Returns TRUE if double x is close to integer.

#' @rdname Theoretical
#' @return \code{dB(x)} returns \code{10*log10(x)}
#' @examples
#' dB( 10) # == 10
#' @export
dB <- function( x) 10.0 * log10( x)

#' @rdname Theoretical
#' @return \code{undB(x)} returns \code{10^(x/10)}
#' @examples
#' undB( 20) # == 100
#' @export
undB <- function( x) 10^( 0.1 * x)

#' @rdname Theoretical
#' @export
Q_ <- function( x) stats::pnorm( x, lower.tail=FALSE)

#' @rdname Theoretical
#' @return \code{Q_Inv(x)} returns \code{2*dB( -qnorm(x))}, which is the
#' SNR (in Decibels) required to get a probability of error of x.
#' Q_Inv( Q_( undB( x/2))) = x and Q_( undB( Q_Inv( x)/2))=x
#'
#' @examples
#' Q_Inv( Q_( undB( 10/2))) # = 10
#' Q_( undB( Q_Inv( 0.001)/2)) # = 0.001
#'
#' @export
Q_Inv <- function( perr) 2.0 * dB( -stats::qnorm( perr))

#' @rdname Theoretical
#'
#' @references \url{https://cran.r-project.org/package=lmomco}
#' @references \url{https://cran.r-project.org/package=gsignal}
#' @export
marcumq <- function( a, b, m=1) {
   stopifnot( a >= 0, b >= 0, m >= 0)
   stats::pchisq(b^2, df = 2 * m, ncp = a^2, lower.tail = FALSE)
}

#' @rdname Theoretical
#' @export
QPSKdB <- function( x) Q_( sqrt_2 * undB( 0.5 * x))

#' @rdname Theoretical
#' @export
DBPSKdB <- function( x) 0.5 * exp( -undB( x))

#' @rdname Theoretical
#' @export
DQPSKdB <- function( x) {
   tgam <- 2 * undB( x)
   oos2 <- sqrt( 0.5)
   a <- sqrt( tgam * (1 - oos2))
   b <- sqrt( tgam * (1 + oos2))
   marcumq( a, b) - 0.5 * besselI( a * b, 0) * exp( -0.5 * (a^2 + b^2))
}

# Differentially decoded DQPSK.
#' @rdname Theoretical
#' @export
DQPSKDDdB <- function( x) {p <- QPSKdB( x); 2.0 * p * (1.0 - p)}

# Polarization shifted QPSK. Aggrell & Karlson
# (2009, DOI:10.1109/JLT.2009.2029064)
#' @rdname Theoretical
#' @export
PSQPSKdB <- function( x) {
   p <- Q_( sqrt( 3.0 * undB( x))); (7.0 / 6.0) * (2.0 - p) * p}

#' @rdname Theoretical
#' @export
MPSKdB <- function( x, M){
   stopifnot( is.wholenumber( M))
   stopifnot( M > 4) # 4 is QPSK, 2 is BPSK
   k <- log2( M);
   (2.0 / k) * Q_( sqrt( 2.0 * k * undB( x)) * sin( pi / M))
}

#' @rdname Theoretical
#' @export
MPSKdB.8 <- function (x) MPSKdB( x, 8)

#' @rdname Theoretical
#' @export
QAMdB.8.star <- function( x) 1.25 * Q_( sqrt( tms3_ * undB( x)))

#' @rdname Theoretical
#' @export
QAMdB <- function( x, M) { # M-ary QAM
   stopifnot( is.wholenumber( M))
   stopifnot( M > 4)
   k <- log2( M)
   (4 / k) * Q_( sqrt( 3.0 * k * undB(x) / (M - 1)))
}

#' @rdname Theoretical
#' @export
QAMdB.16 <- function( x) (1.0 - (1.0 - 1.5 * Q_( sqrt( 0.8 * undB( x))))^2) / 4
#Proakis Eq. 5-2-80

#' @rdname Theoretical
#' @returns \code{ mod_Inv( f, x)} returns a list with the SNR in Decibels to
#' reach the BER
#' \code{perr} such that \code{f( mod_Inv( f, x)$x) = x}.
#' The returned list has elements
#'  \code{$x} as the SNR and
#'  \code{$fval} as the function value.
#'
#' @examples
#' mod_Inv( QPSKdB, QPSKdB( 7)) # yields 7
#'
#' @seealso [pracma::fzero()]
#' @export
## mod_Inv will take a function f(x) and return the x such that f(x)==perr
## but it does this based on the log(f(x)) and the log(perr), so f(x)>0.
mod_Inv <- function( f, perr, guess=Q_Inv( perr))
   if (perr>0)  {pracma::fzero( function( X) log( f( X)) - log( perr), guess)
   } else list( x=Inf, fval=0)

#' @rdname Theoretical
#' @examples
#' mod_InvV(QPSKdB, QPSKdB(c(6,7)))
#'
#' @export
mod_InvV <- function( f, pv, offset=0.0)
   # vectorized version of mod_Inv. pv can be a vector of BERs.
   # The offset is an offset to the Q_ function (in dB), such that
   # the guesses passed to fzero will be Q_Inv( p) - offset.
   apply( cbind( pv, Q_Inv( pv) - offset), 1,
          function( x) mod_Inv( f, x[ 1], x[ 2])$x)


R/B2BConvert.R
#' B2BConvert Converts a function of SNR into one of SNR, B2B, and Offset.
#'
#' Creates a function `f( -dB( undB( -s) + undB( -B2B)) - offset)`
#'
#' Note that all quantities are assumed to be in Decibels.
#'
#' @param f A function of a single argument `f( s)`. `f` can be a symbol, a
#'     string, or an anonymous function.  If a symbol, the symbol name will
#'     be remembered.  If a string, it will be passed to `match.fun` which
#'     will return the function form that is the value of the named symbol
#'     (i.e., not the symbol the string named).
#'
#' @return A function of three arguments `function( x, B2B, offset){...}`
#'     where `x` is the symbol SNR, `B2B` is the back-to-back SNR (i.e. the
#'     equivalent SNR when the input SNR (`x`) is infinite), and `offset` is
#'     the offset to the function `f` (i.e., if `B2B` where infinite).
#' @export
#'
#' @examples
#' QPSKdB.B2B <- B2BConvert( QPSKdB)
#'
B2BConvert <- function( f) {
   force( f)
   fs <- substitute( f) # Try to retain original function name
   if( is.character( fs)) fs <- match.fun( fs)
   # bquote returns a call, and call's remember their source,
   # that is, the source *before* bquote is applied.
   # eval will return a function, but it will attach the call's
   # source.  removeSource clears this, and the source can be
   # recovered from the tokenized code.
   utils::removeSource( eval( bquote( function( x, B2B = Inf, offset = 0)
      .(fs)( -dB((undB( -x) + undB( -B2B))) - offset))))
}


R/cor.R
#' cor.mle returns the correlation matrix of an mle fit.
#'
#' This simple function will return the correlation matrix for `mle` fits of
#' more than one variable.  Note that this function is *not* an S3 generic.
#'
#' @param m an `mle` object with a fit of more than one variable.
#'
#' @return a symmetric matrix with ones on the diagonal.
#'
#' @seealso [stats4::mle()]
#'
#' @export
#'
#' @examples
#' Q_ab <- function( gamma, a, b)
#` Q_(sqrt( a * gamma / (1 + b * gamma)))
#` nobs <- 6 # This is the Mathematica test data.
#` s <- c( sqrt(10), 10, 10 * sqrt(10), 100, 100 * sqrt(10), 1000)
#` r <- c( 246, 211, 220, 23, 9, 6)
#` N <- c( 20000, 500000, 20000000, 20000000, 20000000, 20000000)
#` mle1 <-  mleB2B( Errors=r, N=N, f=Q_ab, fparms=list( gamma=s),
#`               start=c(a=2, b=undB(-11)), method="L-BFGS-B", lower=list(b=0))
#' cor.mle( mle1)
#'
cor.mle <- function( m) {
   stopifnot( isa( m, "mle"))
   c <- stats4::vcov( m)
   v<- diag(1 / sqrt( diag( c)))
   v %*% c %*% v
}
